### ISSUE TITLE
(codepipeline): Error: The 'account' property must be a concrete value (action: 'CodeBuild')

### ISSUE THREAD

**Issue by callumthomson** [aws/aws-cdk#14741]
I have a CodePipeline pipeline with a GitHub source stage, a build stage which builds a Docker image using CodeBuild, and a deploy stage which deploys the image to ECS Fargate. I've been using the CDK to **successfully construct quite a few of these similar pipelines recently however I'm now stuck on 1.99.0 as this error started happening in 1.100.0** and I've tried each release since then up to the most recent release (1.104.0) and they're all not working. I've had a look at the release notes from 1.100.0 and have had a fiddle but have been unable to figure out what's going wrong.

My code fails to validate. For example, when I run `npx cdk diff --all`, I get this error:

```
C:\path\to\my\project\node_modules\@aws-cdk\aws-codepipeline\lib\pipeline.ts:679
      // For every output artifact, get the producer
              ^
Error: The 'account' property must be a concrete value (action: 'CodeBuild')
```

### Reproduction Steps
I've written a reusable function which takes props and creates a CodePipeline pipeline so I'm not sure exactly how to provide minimal reproduction code but here is the function. I'm using it to construct all my deployment pipelines in this project:

```typescript
export interface BuildDeploymentPipelineProps {
  scope: cdk.Construct,
  id: string,
  envName: string,
  name: string,
  artifactBucket: s3.Bucket,
  gitHub: {
    owner: string,
    repo: string,
    branch: string,
    connectionArn: string,
  },
  codebuildProject: codebuild.Project,
  additionalBuildEnv?: { [name: string]: codebuild.BuildEnvironmentVariable },
  deploymentEcsServices: DeploymentEcsService[],
}

export interface DeploymentEcsService {
  name: string,
  service: ecs.FargateService
}

export const buildDeploymentPipeline = (props: BuildDeploymentPipelineProps): codepipeline.Pipeline => {
  const sourceOutput = new codepipeline.Artifact();
  const buildOutput = new codepipeline.Artifact();
  const deployActions: codepipeline.IAction[] = [];
  props.deploymentEcsServices.forEach((deployService) => {
    deployActions.push(
      new cpActions.EcsDeployAction({
        actionName: 'EcsDeploy' + deployService.name,
        service: deployService.service,
        input: buildOutput
      })
    )
  })
  return new codepipeline.Pipeline(props.scope, props.id, {
    pipelineName: naming.constructCodePipelinePipelineName(props.envName, props.name),
    artifactBucket: props.artifactBucket,
    restartExecutionOnUpdate: false,
    crossAccountKeys: false,
    stages: [
      {
        stageName: 'Source',
        actions: [
          new cpActions.CodeStarConnectionsSourceAction({
            actionName: 'GitHubSource',
            owner: props.gitHub.owner,
            repo: props.gitHub.repo,
            connectionArn: props.gitHub.connectionArn,
            output: sourceOutput,
            branch: props.gitHub.branch,
          })
        ]
      },
      {
        stageName: 'Build',
        actions: [
          new cpActions.CodeBuildAction({
            actionName: 'CodeBuild',
            project: codebuild.Project.fromProjectArn(props.scope, 'CodeBuildProject', props.codebuildProject.projectArn),
            input: sourceOutput,
            outputs: [buildOutput],
            environmentVariables: {
              ...props.additionalBuildEnv,
              IMAGE_TAG: {
                type:  codebuild.BuildEnvironmentVariableType.PLAINTEXT,
                value: props.gitHub.branch
              }
            },
          })
        ]
      },
      {
        stageName: 'Deploy',
        actions: deployActions
      }
    ]
  });
}
```

### What did you expect to happen?
Exactly as I'm experiencing in **1.99.0**, the pipeline is created.

### What actually happened?
From **1.100.0** onwards, I'm getting the following error:

```
C:\path\to\my\project\node_modules\@aws-cdk\aws-codepipeline\lib\pipeline.ts:679
      // For every output artifact, get the producer
              ^
Error: The 'account' property must be a concrete value (action: 'CodeBuild')
    at Pipeline.getOtherStackIfActionIsCrossAccount (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\pipeline.ts:679:15)
    at Pipeline.getRoleFromActionPropsOrGenerateIfCrossAccount (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\pipeline.ts:637:36)
    at Pipeline.getRoleForAction (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\pipeline.ts:574:27)
    at Pipeline._attachActionToPipeline (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\pipeline.ts:405:29)
    at Stage.attachActionToPipeline (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\private\stage.ts:141:27)
    at Stage.addAction (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\private\stage.ts:91:29)
    at new Stage (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\private\stage.ts:38:12)
    at Pipeline.addStage (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\pipeline.ts:335:19)
    at new Pipeline (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\aws-codepipeline\lib\pipeline.ts:319:12)
    at Object.exports.buildDeploymentPipeline (C:\Repositories\hiwaldo\infrastructure\lib\services\services\buildDeploymentPipeline.ts:45:10)
Subprocess exited with error 1
```

### Environment

  - **CDK CLI Version  :** aws-cli/2.1.10
  - **Framework Version :** 1.100.0
  - **Node.js Version:** 14.15.3
  - **OS               :** Windows 10 Pro
  - **Language (Version):** TypeScript (3.9.9)

### Other

--- 

This is :bug: Bug Report

**Comment by skinny85** [comment]
Hey @callumthomson,

thanks for opening the issue. Can you show me:

1. What values are used in `codebuild.Project.fromProjectArn(props.scope, 'CodeBuildProject', props.codebuildProject.projectArn)`?
2. What value are you using for the `account` property of the Stack the `scope` argument you pass to `buildDeploymentPipeline` belongs to?

Thanks,
Adam

**Comment by callumthomson** [comment]
Hey @skinny85 thanks for taking the time to look into this.

**Question 1**
Whichever stack is calling the `buildDeploymentPipeline` function will pass in itself (`this`) as `props.scope`. As for `props.codebuildProject`, this object is passed down a tree from higher up. Allow me to illustrate our stacks quickly.
- `App`
    - `EcrStack` 
    - `ParameterStoreStack`
    - `CodebuildStack`
    - `OtherStack`
    - `DevStack`
    - These stacks are combined into a variable called `generalStacks` which are passed into each environment stack below.
    - `EnvironmentStack` - 3 instances of this stack (for staging, uat, and prod)
        - `NetworkStack`
        - `StorageStack`
        - `CommunicationStack`
        - `DeliveryStack`
        - `ServicesStack` - `buildDeploymentPipeline` **is called in this stack**.
            - `AmplifyAppStack`
            - `EcsServiceStack` - `buildDeploymentPipeline` **is called in this stack**.

Here is how `buildDeploymentPipeline` is called inside `ServicesStack`:
```typescript
buildDeploymentPipeline({
  scope: this,
  id: 'CodePipelineAppStoresDeploymentPipeline',
  envName: props.envConfig.envName, // e.g. 'staging'
  name: 'app-stores',
  artifactBucket: props.storageStack.codepipelineArtifactsBucket,
  gitHub: {
    owner: props.envConfig.gitRepositoryOwner,
    repo: 'app-store',
    branch: props.envConfig.branchesName, // e.g. 'staging'
    connectionArn: props.envConfig.gitHubConnectionArn,
  },
  codebuildProject: props.generalStacks.codebuildStack.projects.appStoreBuildImage, // This is a project define in `CodebuildStack`
  deploymentEcsServices: storeAppsDeploymentServices,
});
```

Here is how `buildDeploymentPipeline` is called inside `EcsServiceStack`:
```typescript
this.codepipelineDeploymentPipeline = buildDeploymentPipeline({
  scope: this,
  id: 'CodePipelineDeploymentPipeline',
  envName: props.envConfig.envName,
  name: props.config.name,
  artifactBucket: props.s3CodepipelineArtifactsBucket,
  gitHub: {
    owner: props.envConfig.gitRepositoryOwner,
    repo: props.config.name,
    branch: props.envConfig.branchesName,
    connectionArn: props.envConfig.gitHubConnectionArn,
  },
  codebuildProject: props.codebuildBuildProject, // The CodeBuild project in passed in from the parent stack (ServicesStack) e.g. props.generalStacks.codebuildStack.projects.servicePartnersBuildImage
  additionalBuildEnv: {
    ...props.codebuildAdditionalBuildEnv,
    ...props.config.codepipeline.codebuild?.environment
  },
  deploymentEcsServices: [{
    name: 'Service',
    service: this.ecsFargateService
  }],
});
```

**Question 2**
At the root of my project (`cdk.ts`) I'm defining this `env` variable:
```typescript
const env: cdk.Environment = {
  account: '<my AWS account ID>', // Assuming I shouldn't post this here?
  region: 'eu-west-2',
}
```

This is used as the `env` property in the constructor of all stacks defined directly in the root. My understanding is that this value is passed down automatically into nested stacks.


I hope this helps!

**Comment by skinny85** [comment]
1. Could you show me the values that are passed for `props.codebuildProject.projectArn`?
2. So all 3 Stacks live in the same account and region?

**Comment by skinny85** [comment]
Sorry. Are `NetworkStack`, `StorageStack`, `CommunicationStack`, `DeliveryStack`, `ServicesStack`, `AmplifyAppStack`, `EcsServiceStack` - are they all nested Stacks?

**Comment by skinny85** [comment]
As in, instances of the [`NestedStack` class](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_core.NestedStack.html)?

**Comment by callumthomson** [comment]
The values are shown as tokens, when I console log them. Is there a way to find out what their values are?

Correct, everything here is running in the same account and same region.
Exactly right - in my above list, all direct children of `App` extend `cdk.Stack`, and everything else extends `cdk.NestedStack`

Thanks

**Comment by skinny85** [comment]
> The values are shown as tokens, when I console log them. Is there a way to find out what their values are?

Ah, I see. So `props. codebuildProject` is an actual Project instance? If so, why are you using `codebuild.Project.fromProjectArn(()`? Why not just use the `props.codebuildProject` directly?

**Comment by callumthomson** [comment]
If I substitute the line:
```typescript
//...
project: codebuild.Project.fromProjectArn(props.scope, 'CodeBuildProject', props.codebuildProject.projectArn),
//...
```
for
```typescript
//...
project: props.codebuildProject,
//...
```

I get the following error:
```
Error: 'GeneralCodeBuildStack' depends on 'StagingStack' (GeneralCodeBuildStack -> StagingStack/StorageStack.NestedStack/StorageStack.NestedStackResource.Outputs.${Token[StagingStack.StorageStack.StagingStackStorageStackS3BucketCodepipelineArtifacts12932A47Arn.LogicalID.3608]}). Adding this dependency (StagingStack/ServicesStack/AppDoctorEcsServiceStack -> GeneralCodeBuildStack/CodeBuildProjectBuildImageAppDoctor/Resource.Ref) would create a cyclic reference.
    at EnvironmentStack._addAssemblyDependency (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\stack.ts:732:13)
    at Object.addDependency (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\deps.ts:52:20)
    at EcsServiceStack.addDependency (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\stack.ts:487:5)
    at resolveValue (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\private\refs.ts:100:12)
    at Object.resolveReferences (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\private\refs.ts:30:24)
    at Object.prepareApp (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\private\prepare-app.ts:31:3)
    at Object.synthesize (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\private\synthesis.ts:24:3)
    at App.synth (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\stage.ts:188:23)
    at process.<anonymous> (C:\Repositories\hiwaldo\infrastructure\node_modules\@aws-cdk\core\lib\app.ts:123:45)
    at Object.onceWrapper (events.js:422:26)
Subprocess exited with error 1
```

**Comment by skinny85** [comment]
Right. That means your current Stack split will not work 🙂.

Are you saying that, before, these CodePipelines would execute successfully (I understand CDK succeeded - I'm asking whether the CodePipelines that were produced from this were green)?

**Comment by callumthomson** [comment]
Yeah, with CDK 1.99.0, my pipelines are created and are successfully taking code from GitHub, building images, pushing to ECR and deploying to ECS. 100% as expected. With CD 1.100.0 and above, I get the error during synthesis

**Comment by skinny85** [comment]
Ok. Try this:

```ts
      {
        stageName: 'Build',
        actions: [
          new cpActions.CodeBuildAction({
            actionName: 'CodeBuild',
            project: codebuild.Project.fromProjectName(props.scope, 'CodeBuildProject', props.codebuildProject.projectName),
           // ...
```

And see if that helps.

**Comment by callumthomson** [comment]
Awesome well that's now synthesized but I'm getting an error during deployment

```
Embedded stack arn:aws:cloudformation:eu-west-2:155141762784:stack/staging-stack-ServicesStackNestedStackServicesStackNestedStackResourceD660EA4B-K3B5VXAGDWIR/5ffa6ff0-9de2-11eb-941a-068d6536bcf8 was not successfully updated. Currently in UPDATE_ROLLBACK_IN_PROGRESS with reason: No export named general-codebuild-stack:ExportsOutputRefCodeBuildProjectBuildImage760C360764136162 found
```

Is this something to do with the fact that I'm trying to update an already existing stack or is it something else?

Thanks!

**Comment by skinny85** [comment]
During deployment of what, though?

**Comment by callumthomson** [comment]
Previously `npx cdk synth` (or any command which required synthesis) was throwing the error, now that command is executing successfully.

However I've just tried running `npx cdk deploy -e StagingStack` to try updating our staging stack and it runs successfully for a short while and but then fails because of the above error.

**Comment by skinny85** [comment]
Of course, because you also need to deploy the `CodebuildStack`. Basically, remove the `-e` switch from that above command.

**Comment by callumthomson** [comment]
Ah yes, of course. I'm getting another error now 😄 `npx cdk deploy StagingStack` gives me the following error

```
Export general-codebuild-stack:ExportsOutputFnGetAttCodeBuildProjectBuildImageAppDoctor857DF503Arn7E0CFCCF cannot be deleted as it is in use by prod-stack-ServicesStackNestedStackServices-AppDoctorEcsServiceStackNestedStackAppDoct-1TDVV5TK7Y6TT, staging-stack-ServicesStackNestedStackServi-AppDoctorEcsServiceStackNestedStackAppDoct-HZY70HQXG055 and uat-stack-ServicesStackNestedStackServicesS-AppDoctorEcsServiceStackNestedStackAppDoct-PXE8YMANIALD
```

I'm not sure why it's trying to delete the export from `GeneralCodeBuildStack` - I must admit my CloudFormation / CDK understnanding is a bit vague in this area. This must be because the stacks have already been deployed using `fromProjectArn` and cannot be just simply updated to `fromProjectName` without some trickery, right?

**Comment by skinny85** [comment]
Yes, you're exactly right, but there's a simple trick you can employ to keep the existing export. Add this to `buildDeploymentPipeline`:

```ts
  new CfnOutput(props.scope, 'ProjectArnOutput', {
    value: props.codebuildProject.projectArn,
  });
```

I _believe_ this should make the deploy succeed.

**Comment by callumthomson** [comment]
Awesome! That did work! I did run into another issue (I solved it though) after I had finished deploying all stacks. I didn't want to have that snippet of code just lying around forever so I removed it and I got a similar error to before:

```
Export general-codebuild-stack:<whatever> cannot be deleted as it is in use by prod-stack-<whatever> and uat-stack-<whatever>
```

I couldn't understand why this was the case, but then I realised that the CDK was probably trying to update `GeneralCodeBuildStack` to remove the export *before* it updated `StagingStack`, `UatStack` and `ProdStack` to not require the export. To resolve this problem I just ran the following commands:

```
$ npx cdk deploy -e StagingStack
$ npx cdk deploy -e UatStack
$ npx cdk deploy -e ProdStack
$ npx cdk deploy -e GeneralCodeBuildStack
```

Exclusively deploying the others stacks first, so that they are no longer are importing the export from the `GeneralCodeBuildStack`, meant that I could then deploy `GeneralCodeBuildStack` and not have CloudFormation complain that exports were in use by other stacks. But this got me thinking: would it be possible for the CDK to identify such cases so that it would automatically deploy the dependent stacks first? Come to think of it, this has actually happened to me before and I've resolved it in the same way. It's easy for me to see that certain stacks should deploy first, but I must manually exclusively deploy them in the correct order to get it to work.

So finally then, I've upgraded to CDK 1.105.0 everything is working fine. Thanks so much for your help! But now I must ask - is this a bug? Using the project name instead of the project ARN, while it works, seems less robust. I certainly can't understand why `fromProjectName` works and `fromProjectArn` doesn't - especially when it stopped working so suddenly like that.

Thanks

**Comment by skinny85** [comment]
Nice, I'm glad you got it working.

As far as if this is a bug - I don't think so. So basically, the situation is that your Pipeline is in a specific account (that you provided by passing it in the `env` property of the `Stack` it belongs to). When you did this:

```ts
project: codebuild.Project.fromProjectArn(props.scope, 'CodeBuildProject',
  props.codebuildProject.projectArn),
```

You basically "erased" the account of the Project by passing this for the ARN (because it's an opaque `Fn::GetAtt` expression now, and we can't know what account that ARN belongs to).

Using this instead:

```ts
project: codebuild.Project.fromProjectName(props.scope, 'CodeBuildProject',
  props.codebuildProject.projectName),
```

you're telling us "the Project belongs to the same account (and region) as `props.scope`", and so we can safely compare it to the account of the Pipeline, and confirm they are in the same account.

Hope this clears this up!

**Comment by callumthomson** [comment]
Ok, I understand but, how come it was working fine before in 1.99.0? I see this in the release notes for 1.100.0:

> codepipeline: incorrect determination of the Action's account when using an imported resource (#14224) (d88e915), closes #14165

This sounds extremely related to the issue I was facing, and does the ARN not contain the account number?

But I acknowledge I'm not familiar enough with the CDK / CloudFormation to know for sure, so if you say it's expected behaviour, then I will accept that.

**Comment by skinny85** [comment]
You got it right again 🙂. The PR in #14224 fixed a bug (#14165), but it also accidentally surfaced this problem in your code.

But I would argue the code is actually correct now, and allowing the `fromProjectArn` to work was a mistake from the beginning.

**Comment by callumthomson** [comment]
Ok, understood. Let's consider this problem solved then 😄. Thanks again for your investigation and assistance.

Callum
